File: src/my_py_pkg/my_py_pkg/led_panel_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from my_robot_interfaces.msg import LedStates
from my_robot_interfaces.srv import SetLed

class LedPanelNode(Node):

    def __init__(self):
        super().__init__("led_panel") 
        
        self.declare_parameter("led_states", [0, 0, 0])
        self.led_states_ = self.get_parameter("led_states").value

        self.led_states_publisher_ = self.create_publisher(LedStates, "led_states", 10) # create topic led_states in which the node will publish
        self.timer = self.create_timer(1.0, self.publish_led_states) # publish 2 times per sec 
        
        self.set_led_server_ = self.create_service(SetLed, "set_led", self.callback_set_led)

        self.get_logger().info("Led Node has been started")
    
    def publish_led_states(self):
        msg = LedStates()
        msg.led_states = self.led_states_
        self.led_states_publisher_.publish(msg)
    
    def callback_set_led(self, request, response):
        if(request.led_num >= 0 and request.led_num < len(self.led_states_)):
            self.led_states_[request.led_num] = request.state
            response.success = True
        else:
            response.success = False
        return response
    

def main(args=None):
    rclpy.init(args=args) # first line to be written in any ROS2 .py node
    node = LedPanelNode()
    rclpy.spin(node) # allow node to continue to be alive
    rclpy.shutdown() # last line of any ROS2 .py node

if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/led_panel_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/led_states.hpp"
#include "my_robot_interfaces/srv/set_led.hpp"

using std::placeholders::_1;
using std::placeholders::_2;
using std::vector;
using std::chrono::seconds;

using my_robot_interfaces::msg::LedStates;
using my_robot_interfaces::srv::SetLed;

class LedPanelNode : public rclcpp::Node
{
public:
    LedPanelNode() : Node("led_panel")
    {
        this->declare_parameter("led_states", vector<int64_t> (3, 0));
        this->led_states_ = this->get_parameter("led_states").as_integer_array();
        led_states_publisher_ = this->create_publisher<LedStates>("led_states", 10);
        timer_ = this->create_wall_timer(seconds(1),
                                         bind(&LedPanelNode::publishLedStates, this));

        set_led_server_ = this->create_service<SetLed>("set_led", std::bind(&LedPanelNode::callbackSetLed, this, _1, _2));

        RCLCPP_INFO(this->get_logger(), "Led Panel Node has been started");
    }

private:
    void publishLedStates()
    {
        auto msg = LedStates();
        msg.led_states = this->led_states_;
        led_states_publisher_->publish(msg);
    }

    void callbackSetLed(const SetLed::Request::SharedPtr request,const SetLed::Response::SharedPtr response)
    {
        if(request->led_num >= 0 && request->led_num <= (int)this->led_states_.size()){
            this->led_states_[request->led_num] = request->state;
            response->success = true;
        }else
            response->success = false;
    }

    rclcpp::Publisher<LedStates>::SharedPtr led_states_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Service<SetLed>::SharedPtr set_led_server_;
    vector<int64_t> led_states_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    //shared pointer, you don't need to delete manually the instance when object is not referenced (like GC in Java)
    auto node = std::make_shared<LedPanelNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}